### ISSUE TITLE
Venn Diagram: Set input names

### ISSUE THREAD

**Issue by ajdapretnar** [biolab/orange3#1083]
Input names displayed in Venn Diagram are often strange. If an input comes from Select Rows widget, the input name in Venn should say this (e.g. "Fruits_and_Vegetables.xlsx - subset 1"). Also allow for a custom rename of inputs directly in Venn.

**Comment by kernc** [comment]
@BlazZupan, @janezd How does a suffix like "subset 1" .. "subset n" help in discerning where the input came from? (In a reloaded workspace, for example.)

**Comment by ajdapretnar** [comment]
This one is for me to explain. Subset 1 obviously isn't a great way to represent the subset, but now it's even worse with the entire data set name displayed in Venn (eg. 'iris' and 'iris'). Is it possible to come up with a more descriptive and/or recognizable input names?

Also I'm having trouble with Select Rows input. Venn displays nothing to two Select Rows inputs, while Data Table normally contains the subset. Try with Iris and use Select Rows (sepal length greater than 5) and Select Rows (petal length greater than 3) to reproduce.

**Comment by kumargaurav2596** [comment]
Can you please explain in little detail, the naming scheme that you are hoping to implement.

Beside I think their is another issue that, the venn diagram show 0 items even though the set is non empty. Is that an issue or I am missing on something.

![venn_diagram](https://cloud.githubusercontent.com/assets/10504619/13783019/9d01167a-eaef-11e5-85a5-9e1caa4aa081.png)

**Comment by kernc** [comment]
@kumargaurav2596 That's the issue https://github.com/biolab/orange3/issues/1099 and its fix https://github.com/biolab/orange3/pull/1110. You can verify that [patch](https://github.com/biolab/orange3/pull/1110.patch) works for you.

**Comment by kumargaurav2596** [comment]
Should I include double click on level to rename it functionality.

**Comment by ajdapretnar** [comment]
@kernc @kumargaurav2596 Could we just add the option for the user to rename the input in Venn and leave the default as is?

**Comment by kernc** [comment]
@ajdapretnar Add the option _how_?

@kumargaurav2596 Doubleclick sounds like it will be hidden and overlooked (i.e. no one will know about it). Probably better if an [edit icon](https://www.google.com/search?q=edit+icon&tbm=isch) pops up next to it on hover.

**Comment by ajdapretnar** [comment]
> @ajdapretnar Add the option how?

I guess the way it's implemented in Color widget. But I also like the idea of the edit icon on hover.

**Comment by kumargaurav2596** [comment]
@kernc  doubleclick to edit is already available, how about on hover we just display "double click to edit" message.

**Comment by ajdapretnar** [comment]
Should be fixed via #1481.

**Comment by kernc** [comment]
I don't think that PR has anything to do with this issue. Is this solvable even in concept? https://github.com/biolab/orange3/issues/1083#issuecomment-194884890

**Comment by ajdapretnar** [comment]
Disagree. The PR should go back to *iris* and *iris* being displayed in Venn, which is much better than 'untitled'. @lanzagar Please confirm. 

One can already manually rename the input, should she wish to do so. That's about as much as we can do about this.

**Comment by kernc** [comment]
Yes, #1481 solves the untitled thing. I thought this issue was about the user not knowing which set in the diagram belongs to which input source. This issue can also be about lack of proper indicators that the labels can indeed be edited. :smiley:

**Comment by ajdapretnar** [comment]
> I thought this issue was about the user not knowing which set in the diagram belongs to which input source.

Yes, this was originally a part of the issue, but after discussing this at length, I see no better solution to this other than manually renaming the input. However, we might still consider adding canvas widget name after the data set name (e.g. iris (Select Rows (1)). Is that possible?

Agree, editing indicators should be fixed.

**Comment by kernc** [comment]
> adding canvas widget name after the data set name (e.g. iris (Select Rows (1))

:+1:

**Comment by kernc** [comment]
> adding canvas widget name after the data set name (e.g. iris (Select Rows (1))

Unfortunately, this is not possible. Widgets have no reference to canvas or their previous connections.

**Comment by ajdapretnar** [comment]
I thought that's the case. We've seriously discussed this at length and there are few possible solutions. :disappointed:

**Comment by lanzagar** [comment]
As far as I can tell, this can be closed.
Editing names worked already, and after #1481 the default should also be fixed (tables have correct names instead of 'untitled').

The only other improvement that makes sense to me would be to make the changed names a schema_only setting.

**Comment by ajdapretnar** [comment]
As @kernc mentioned, the true issue here was of a different kind. The problem is that if you, for example, connect two Select Rows widgets to Venn, both sets would have the same name (i.e. 'iris'). So you cannot really know which set belongs to which input. Naturally, you can go back to Select Rows and probably count the no of instances on the output, but that's not really practical.

Can anyone think of a way how to elegantly label sets in Venn according to the input (selected subset from Scatter Plot, Select Rows, two File widgets, etc.)?

If not, I vote for closing the issue.

**Comment by astaric** [comment]
Widgets do not know where the input is coming from (this is by design). Unless we decide to append widget name in data that is sent **from the widget**, there is nothing we can do to differentiate the inputs. If we decide to do that (at a later date), this will include modifying all the widgets that output data and has nothing to do with Venn Diagram.

The rest of the problems mentioned in this issue have already been fixed (datasets named untitled, editable names, visual indicator that the names can be edited).
